File: src/main/java/de/thelooter/eventchecker/EventChecker.java
package de.thelooter.eventchecker;

import de.thelooter.eventchecker.commands.EventCheckerCommand;
import de.thelooter.eventchecker.commands.complete.EventCheckerCommandCompleter;
import io.github.classgraph.ClassGraph;
import io.github.classgraph.ClassInfo;
import io.github.classgraph.ClassInfoList;
import org.apache.commons.lang3.exception.ExceptionUtils;
import org.bukkit.Bukkit;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.event.Event;
import org.bukkit.event.EventPriority;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.plugin.EventExecutor;
import org.bukkit.plugin.PluginDescriptionFile;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.plugin.java.JavaPluginLoader;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class EventChecker extends JavaPlugin implements Listener {

  public static List<String> eventNames = new ArrayList<>();

  @Override
  public void onEnable() {
    ClassInfoList events =
        new ClassGraph()
            .enableClassInfo()
            .scan()
            .getClassInfo(Event.class.getName())
            .getSubclasses()
            .filter(info -> !info.isAbstract());

    events.forEach(event -> eventNames.add(event.getName()));

    File configFile = new File(getDataFolder(), "config.yml");
    if (!configFile.exists()) {
      configFile.getParentFile().mkdirs();
      saveDefaultConfig();
    }

    FileConfiguration config = new YamlConfiguration();

    try {
      config.load(configFile);
    } catch (Exception e) {
      getLogger().severe("Could not load config file: " + ExceptionUtils.getStackTrace(e));
    }

    boolean blacklist = config.getBoolean("blacklist", false);
    List<String> excludedEvents = config.getStringList("excluded-events");

    boolean whitelist = config.getBoolean("whitelist", false);
    List<String> includedEvents = config.getStringList("included-events");

    if (blacklist && whitelist) {
      getLogger().severe("Cannot use both blacklist and whitelist at the same time");
      Bukkit.getPluginManager().disablePlugin(this);
      return;
    }

    Listener listener = new Listener() {};

    EventExecutor executor = (ignored, event) -> getLogger().info("Event: " + event.getEventName());

    List<ClassInfo> enabledEventsBlackList =
        events.stream()
            .filter(
                info -> {
                  for (String excludedEvent : excludedEvents) {
                    if (info.getName()
                        .substring(info.getName().lastIndexOf('.') + 1)
                        .equals(excludedEvent)) {
                      return false;
                    }
                  }
                  return true;
                })
            .toList();

    List<ClassInfo> enabledEventsWhiteList =
        events.stream()
            .filter(
                info -> {
                  for (String includedEvent : includedEvents) {
                    if (info.getName()
                        .substring(info.getName().lastIndexOf('.') + 1)
                        .equals(includedEvent)) {
                      return true;
                    }
                  }
                  return false;
                })
            .toList();

    if (blacklist) {
      registerEvents(listener, executor, enabledEventsBlackList);
    } else if (whitelist) {
      registerEvents(listener, executor, enabledEventsWhiteList);
    } else {
      registerEvents(listener, executor, events);
    }

    String[] eventNames =
        events.stream()
            .map(
                classInfo ->
                    classInfo.getName().substring(classInfo.getName().lastIndexOf('.') + 1))
            .toArray(String[]::new);

    getLogger().info("List of Events: " + String.join(", ", eventNames));
    getLogger().info("Number of Events: " + events.size());
    getLogger().info("HandlerList Size:" + HandlerList.getHandlerLists().size());

    // Register Commands
    Objects.requireNonNull(getCommand("eventchecker")).setExecutor(new EventCheckerCommand());

    // Register CommandCompletions
    Objects.requireNonNull(getCommand("eventchecker")).setTabCompleter(new EventCheckerCommandCompleter());
  }

  private void registerEvents(
      Listener listener, EventExecutor executor, List<ClassInfo> enabledEvent) {
    for (ClassInfo info : enabledEvent) {
      try {
        @SuppressWarnings("unchecked")
        Class<? extends Event> eventClass = (Class<? extends Event>) Class.forName(info.getName());

        if (Arrays.stream(eventClass.getDeclaredMethods())
            .anyMatch(
                method ->
                    method.getParameterCount() == 0 && method.getName().equals("getHandlers"))) {
          getServer()
              .getPluginManager()
              .registerEvent(eventClass, listener, EventPriority.NORMAL, executor, this);
        }
      } catch (ClassNotFoundException e) {
        throw new RuntimeException(e);
      }
    }
  }
}

File: src/main/kotlin/de/thelooter/eventchecker/EventChecker.kt
package de.thelooter.eventchecker

import de.thelooter.eventchecker.commands.EventCheckerCommand
import de.thelooter.eventchecker.commands.complete.EventCheckerCommandCompleter
import de.thelooter.eventchecker.events.EventTaskManager
import de.thelooter.eventchecker.events.tasks.EventRegistrationTask
import io.github.classgraph.ClassGraph
import io.github.classgraph.ClassInfo
import org.apache.commons.lang3.exception.ExceptionUtils
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.event.Event
import org.bukkit.event.HandlerList
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.util.*

open class EventChecker : JavaPlugin() {

    companion object {
        var eventNames: MutableList<String> = ArrayList()
        lateinit var instance: JavaPlugin
        val eventTaskManager = EventTaskManager()
    }

    /**
     * This method is the onEnable() method of the EventChecker class. It is called when the plugin is enabled.
     * It performs various setup and initialization tasks.
     */
    override fun onEnable() {
        instance = this
        val events = ClassGraph()
            .enableClassInfo()
            .scan()
            .getClassInfo(Event::class.java.name)
            .subclasses
            .filter { info: ClassInfo -> !info.isAbstract }

        events.forEach {
            eventNames.add(it.name)
        }

        val configFile = File(dataFolder, "config.yml").apply {
            if (!exists()) {
                parentFile.mkdirs()
                saveDefaultConfig()
            }
        }


        val config = YamlConfiguration().apply {
            try {
                load(configFile)
            } catch (e: Exception) {
                logger.severe("Could not load config.yml: ${ExceptionUtils.getStackTrace(e)}")
            }
        }


        val blackList = config.getBoolean("blacklist", false)
        val excludedEvents = config.getStringList("excluded-events")

        val whiteList = config.getBoolean("whitelist", false)
        val includedEvents = config.getStringList("included-events")

        if (blackList && whiteList) {
            logger.severe("Both blacklist and whitelist are enabled. Please disable one of them.")
            return
        }

        val enabledEventsBlackList: List<ClassInfo> = events.filter {
            !excludedEvents.contains(it.name.substring(it.name.lastIndexOf(".") + 1))
        }

        val enabledEventsWhiteList: List<ClassInfo> = events.filter {
            includedEvents.contains(it.name.substring(it.name.lastIndexOf(".") + 1))
        }

        if (blackList) {
            this registerEvents enabledEventsBlackList
        } else if (whiteList) {
            this registerEvents enabledEventsWhiteList
        } else {
            this registerEvents events
        }

        logger.info("HandlerList size: " + HandlerList.getHandlerLists().size)

        getCommand("eventchecker")?.apply {
            setExecutor(EventCheckerCommand())
            tabCompleter = EventCheckerCommandCompleter()
        }
    }

    /**
     * Registers all events in the given list.
     * @param enabledEvent The list of events to register.
     *
     * @since 1.3.0
     */
    private infix fun registerEvents(enabledEvent: List<ClassInfo>) {
        enabledEvent.forEach {
            eventTaskManager.addTask(EventRegistrationTask(it))
        }

        eventTaskManager.processTasks()

        val eventNames = enabledEvent.map {
            it.name.substring(it.name.lastIndexOf(".") + 1)
        }
        logger.info("Enabled events: " + eventNames.joinToString(", "))
        logger.info("Number of enabled events: " + eventNames.size)
    }

}
